src/routes/conversation/[id]/web-search/+server.ts

```diff
@@ -73,45 +61,60 @@ export async function GET({ params, locals, url }) {
 
 				appendUpdate("Searching Google", [webSearch.searchQuery]);
 				const results = await searchWeb(webSearch.searchQuery);
-
-				let text = "";
 				webSearch.results =
 					(results.organic_results &&
 						results.organic_results.map((el: { link: string }) => el.link)) ??
 					[];
-
-				if (results.answer_box) {
-					// if google returns an answer box, we use it
-					webSearch.answerBox = JSON.stringify(removeLinks(results.answer_box));
-					text = webSearch.answerBox;
-					appendUpdate("Found a Google answer box");
-				} else if (results.knowledge_graph) {
-					// if google returns a knowledge graph, we use it
-					webSearch.knowledgeGraph = JSON.stringify(removeLinks(results.knowledge_graph));
-					text = webSearch.knowledgeGraph;
-					appendUpdate("Found a Google knowledge page");
-				} else if (webSearch.results.length > 0) {
-					let tries = 0;
-
-					while (!text && tries < 3) {
-						const searchUrl = webSearch.results[tries];
-						appendUpdate("Browsing result", [JSON.stringify(searchUrl)]);
+				webSearch.results = webSearch.results
+					.filter((link) => !link.includes("youtube.com")) // filter out youtube links
+					.slice(0, MAX_N_PAGES_SCRAPE); // limit to first 10 links only
+
+				let paragraphChunks: string[] = [];
+				if (webSearch.results.length > 0) {
+					appendUpdate("Browsing results");
+					const promises = webSearch.results.map(async (link) => {
+						let text = "";
 						try {
-							text = await parseWeb(searchUrl);
-							if (!text) throw new Error("text of the webpage is null");
+							text = await parseWeb(link);
 						} catch (e) {
-							appendUpdate("Error parsing webpage", [], "error");
-							tries++;
+							console.error(`Error parsing webpage "${link}"`, e);
+							appendUpdate("Error parsing webpage", [`url: ${link} error: ${e}`], "error");
 						}
+						const CHUNK_CAR_LEN = 512;
+						const MAX_N_CHUNKS = 100;
+						const chunks = chunk(text, CHUNK_CAR_LEN).slice(0, MAX_N_CHUNKS);
+						return chunks;
+					});
+					const nestedParagraphChunks = (await Promise.all(promises)).slice(0, MAX_N_PAGES_EMBED);
+					paragraphChunks = nestedParagraphChunks.flat();
+					if (!paragraphChunks.length) {
+						throw new Error("No text found on the first 5 results");
 					}
-					if (!text) throw new Error("No text found on the first 3 results");
 				} else {
 					throw new Error("No results found for this search query");
 				}
 
-				appendUpdate("Creating summary");
-				webSearch.summary = await summarizeWeb(text, webSearch.searchQuery, model);
-				appendUpdate("Injecting summary", [JSON.stringify(webSearch.summary)]);
+				appendUpdate("Extracing relevant information");
+				const topKClosestParagraphs = 8;
+				const requestBody = {
+					paragraphs: paragraphChunks,
+					query: prompt,
+					top_k: topKClosestParagraphs,
+				};
+				const res = await fetch("https://mishig-embeddings-similarity.hf.space/", {
```

Comment:
Would it be possible to make this an environment variable so users can have their own embeddings API? :eyes:

Reply:
totally! I'm gonna move this one to https://huggingface.co/huggingchat org

Reply:
so what i would suggest is that it should:
- run by default on server-side JS (transformers.js for instance) embedding + pure JS similarity search
- optionally, use your Space as a backend API (to provide better performance? though i'm pretty sure the local perf would be the same)

Reply:
handled in https://github.com/huggingface/chat-ui/pull/427#issuecomment-1717070246